### PR TITLE
osd: metadata device supports udev path

### LINKED ISSUE #10486: metadata device doesn't support `/dev/disk/by-path/xxxx`
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**

**Expected behavior:**

**How to reproduce it (minimal and precise):**

Specify the metadata device as `/dev/disk/by-path/xxx`, rather than `/dev/sdx`.

**File(s) to submit**:

* Cluster CR (custom resource), typically called `cluster.yaml`, if necessary
* Operator's logs, if necessary
* Crashing pod(s) logs, if necessary

 To get logs, use `kubectl -n <namespace> logs <pod name>`
When pasting logs, always surround them with backticks or use the `insert code` button from the Github UI.
Read [GitHub documentation if you need help](https://help.github.com/en/articles/creating-and-highlighting-code-blocks).

**Environment**:
* OS (e.g. from /etc/os-release):
* Kernel (e.g. `uname -a`):
* Cloud provider or hardware configuration:
* Rook version (use `rook version` inside of a Rook Pod):
```
[rook@rook-ceph-operator-66c447c48b-wpkfv /]$ rook version
rook: v1.8.5
go: go1.16.13
```
* Storage backend version (e.g. for ceph do `ceph -v`):
```
ceph version 16.2.7
```
* Kubernetes version (use `kubectl version`):
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift):
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/latest/Troubleshooting/ceph-toolbox/#interactive-toolbox)):

---
Osd devices supports specifying udev paths `/dev/disk/by-path/xxxx`, but metadata devices only support disk symbols `/dev/sdx`. Rook automatically converts udev path corresponding to the OSD into a disk symbol, except for metadata device. It comes to normal after change metadata device from udev path to disk.

ceph cluster snippet:
```
- config:
          metadataDevice: /dev/disk/by-path/pci-0000:d8:00.0-nvme-1
        devices:                                                      
        - name: /dev/disk/by-path/pci-0000:86:00.0-scsi-0:2:34:0
        - name: /dev/disk/by-path/pci-0000:86:00.0-scsi-0:2:22:0
        - name: /dev/disk/by-path/pci-0000:86:00.0-scsi-0:2:7:0
        - name: /dev/disk/by-path/pci-0000:87:00.0-scsi-0:2:30:0
        - name: /dev/disk/by-path/pci-0000:87:00.0-scsi-0:2:19:0
        - name: /dev/disk/by-path/pci-0000:86:00.0-scsi-0:2:19:0
        - name: /dev/disk/by-path/pci-0000:87:00.0-scsi-0:2:27:0
        - name: /dev/disk/by-path/pci-0000:86:00.0-scsi-0:2:4:0
        - name: /dev/disk/by-path/pci-0000:86:00.0-scsi-0:2:42:0
        - name: /dev/disk/by-path/pci-0000:87:00.0-scsi-0:2:28:0
        - name: /dev/disk/by-path/pci-0000:86:00.0-scsi-0:2:15:0
        - name: /dev/disk/by-path/pci-0000:87:00.0-scsi-0:2:46:0
        name: test3
        resources: {}
      useAllDevices: false
```
```
[root@test1 ~]# ll  /dev/disk/by-path/pci-0000:d8:00.0-nvme-1
lrwxrwxrwx 1 root root 13 Jun 22 14:18 /dev/disk/by-path/pci-0000:d8:00.0-nvme-1 -> ../../nvme0n1
```

log snippet:
```
2022-06-22 05:54:15.167841 I | cephosd: configuring new LVM device sdal
2022-06-22 05:54:15.167845 I | cephosd: using /dev/disk/by-path/pci-0000:d9:00.0-nvme-1 as metadataDevice for device /dev/sdal and let ceph-volume lvm batch decide how to create volumes
2022-06-22 05:54:15.167848 I | cephosd: configuring new LVM device sdcd
2022-06-22 05:54:15.167850 I | cephosd: using /dev/disk/by-path/pci-0000:d9:00.0-nvme-1 as metadataDevice for device /dev/sdcd and let ceph-volume lvm batch decide how to create volumes
2022-06-22 05:54:15.167853 I | cephosd: configuring new LVM device nvme0n1
2022-06-22 05:54:15.167856 I | cephosd: using /dev/disk/by-path/pci-0000:d9:00.0-nvme-1 as metadataDevice for device /dev/nvme0n1 and let ceph-volume lvm batch decide how to create volumes
2022-06-22 05:54:15.167859 I | cephosd: configuring new LVM device sdaf
2022-06-22 05:54:15.167862 I | cephosd: using /dev/disk/by-path/pci-0000:d9:00.0-nvme-1 as metadataDevice for device /dev/sdaf and let ceph-volume lvm batch decide how to create volumes
2022-06-22 05:54:15.167867 I | cephosd: configuring new LVM device sdat
2022-06-22 05:54:15.167870 I | cephosd: using /dev/disk/by-path/pci-0000:d9:00.0-nvme-1 as metadataDevice for device /dev/sdat and let ceph-volume lvm batch decide how to create volumes
2022-06-22 05:54:15.167877 I | cephosd: configuring new LVM device sdbc
2022-06-22 05:54:15.167882 I | cephosd: using /dev/disk/by-path/pci-0000:d9:00.0-nvme-1 as metadataDevice for device /dev/sdbc and let ceph-volume lvm batch decide how to create volumes
2022-06-22 05:54:15.167898 D | exec: Running command: stdbuf -oL ceph-volume --log-path /tmp/ceph-log lvm batch --prepare --bluestore --yes --osds-per-device 1 --crush-device-class hdd /dev/sdbt /dev/sdco /dev/sdcr /dev/sdcw /dev/sdd /dev/sdal /dev/sdcd /dev/nvme0n1 /dev/sdaf /dev/sdat /dev/sdbc --db-devices /dev/disk/by-path/pci-0000:d9:00.0-nvme-1 --report
2022-06-22 05:54:20.450931 D | exec: 
2022-06-22 05:54:20.450951 D | exec: Total OSDs: 0
2022-06-22 05:54:20.450954 D | exec: 
2022-06-22 05:54:20.450957 D | exec:   Type            Path                                                    LV Size         % of device
2022-06-22 05:54:20.450968 D | exec: --> passed data devices: 11 physical, 0 LVM
2022-06-22 05:54:20.450981 D | exec: --> relative data size: 1.0
2022-06-22 05:54:20.450986 D | exec: --> passed block_db devices: 1 physical, 0 LVM
2022-06-22 05:54:20.450989 D | exec: --> 1 fast devices were passed, but none are available
2022-06-22 05:54:20.482566 D | exec: Running command: stdbuf -oL ceph-volume --log-path /tmp/ceph-log lvm batch --prepare --bluestore --yes --osds-per-device 1 --crush-device-class hdd /dev/sdbt /dev/sdco /dev/sdcr /dev/sdcw /dev/sdd /dev/sdal /dev/sdcd /dev/nvme0n1 /dev/sdaf /dev/sdat /dev/sdbc --db-devices /dev/disk/by-path/pci-0000:d9:00.0-nvme-1 --report --format json
2022-06-22 05:54:21.889238 D | cephosd: ceph-volume reports: []
2022-06-22 05:54:21.895521 C | rookcmd: failed to configure devices: failed to initialize osd: failed to create enough required devices, required: [], actual: []
```

### PR DESCRIPTION
**Pull request by microyahoo**
Signed-off-by: Liang Zheng <zhengliang0901@gmail.com>

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #10486

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure)).
- [ ] **Skip Tests for Docs**: If this is only a documentation change, add the label `skip-ci` on the PR.
- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

**Review by subhamkrai — CHANGES_REQUESTED** [review]
@microyahoo Thanks for the PR, please look at ci errors.

**Comment by microyahoo** [comment]
@subhamkrai Thanks for your reminder, it is WIP, and I will add some unit tests.

**Comment by subhamkrai** [comment]
> @subhamkrai Thanks for your reminder, it is WIP, and I will add some unit tests.

sure

File: pkg/daemon/ceph/osd/volume.go [hunk 1]
```diff
@@ -641,6 +641,29 @@ func (a *OsdAgent) initializeDevicesLVMMode(context *clusterd.Context, devices *
...
+						logger.Infof("%s found in the desired metadata devices", localDevice.Name)
+						metadataDevice = localDevice
+						break
+					} else if strings.HasPrefix(md, "/dev/") {
+						devLinks := strings.Split(localDevice.DevLinks, " ")
+						for _, link := range devLinks {
+							if link == md {
+								logger.Infof("%s found in the desired metadata device (matched by link: %s)", localDevice.Name, link)
+								metadataDevice = localDevice
+								break OUTER
```
Comment (parth-gr) [inline pkg/daemon/ceph/osd/volume.go:657]: Ig `OUTER` here symbolizes that we need to re-iterate,

I am not sure but this is considered a not a best practice for writing a code(as we never use `goto-statement`),
I am not sure what you finally trying to achieve but if you describe it I can surely say I can suggest you something better.

Maybe @travisn @BlaineEXE @subhamkrai you should also have a look.

Comment (microyahoo) [reply pkg/daemon/ceph/osd/volume.go:657]: Thanks for you suggestion. If we find the corresponding metadata disk, we can break out from the inner loop.

**Comment by BlaineEXE** [comment]
Make sure the commit message is not empty. I see a subject but no body.

File: pkg/daemon/ceph/osd/volume.go (hunk 1 above)
Comment (BlaineEXE) [reply pkg/daemon/ceph/osd/volume.go:657]: Let's not use `go-to` statements like this. It's hard to follow the code, and we have avoided it so far. The better pattern is to make the inner for loop a function of its own. If that function finds a match, then we can just do `break`.

**Review by BlaineEXE — CHANGES_REQUESTED** [review]

File: pkg/daemon/ceph/osd/volume.go [hunk 2]
```diff
@@ -641,6 +641,29 @@ func (a *OsdAgent) initializeDevicesLVMMode(context *clusterd.Context, devices *
...
 					md = device.Config.MetadataDevice
 				}
+				var metadataDevice *sys.LocalDisk
+			OUTER:
+				for _, localDevice := range context.Devices {
+					if localDevice.Name == md || filepath.Join("/dev", localDevice.Name) == md {
+						logger.Infof("%s found in the desired metadata devices", localDevice.Name)
+						metadataDevice = localDevice
+						break
+					} else if strings.HasPrefix(md, "/dev/") {
```
Comment (BlaineEXE) [inline pkg/daemon/ceph/osd/volume.go:651]: Because of this `break` statement. The `else if` below is unnecessary. It would be cleaner as a new `if` statement.

**Review by BlaineEXE — CHANGES_REQUESTED** [review]

**Comment by BlaineEXE** [comment]
I see that the `Canary integration tests / osd-with-metadata-device` test is failing. Is something in this PR breaking old behavior?

**Comment by microyahoo** [comment]
@BlaineEXE I have resolved the comments besides integrated test of [osd-with-metadata-device](https://github.com/rook/rook/runs/7040396677?check_suite_focus=true#logs). I have another questions, why `ceph-volume --report` command runs twice, it seems that run it once should be enough. Any comments is welcome.
https://github.com/rook/rook/blob/c8658331df09cbef3b07b1ac3f60b28267ce1635/pkg/daemon/ceph/osd/volume.go#L752
https://github.com/rook/rook/blob/c8658331df09cbef3b07b1ac3f60b28267ce1635/pkg/daemon/ceph/osd/volume.go#L761

**Comment by microyahoo** [comment]
The integration test uses `lv` as the metadata device. After conversion, `ceph-volume` uses `/dev/dm-0` as the db-device parameter instead of ` /dev/test-rook-vg/test-rook-lv`, resulting in command execution failure.

The following is error logs snippet:
```
2022-06-24 11:10:35.284056 I | cephosd: dm-0 found in the desired metadata device (matched by link: /dev/test-rook-vg/test-rook-lv)
2022-06-24 11:10:35.284148 I | cephosd: using dm-0 as metadataDevice for device /dev/sda and let ceph-volume lvm batch decide how to create volumes
2022-06-24 11:10:35.284240 D | exec: Running command: stdbuf -oL ceph-volume --log-path /tmp/ceph-log lvm batch --prepare --bluestore --yes --osds-per-device 1 --crush-device-class hdd /dev/sda --db-devices /dev/dm-0 --report
2022-06-24 11:10:35.768977 D | exec: --> passed data devices: 1 physical, 0 LVM
2022-06-24 11:10:35.769194 D | exec: --> relative data size: 1.0
2022-06-24 11:10:35.769563 D | exec: --> passed block_db devices: 0 physical, 1 LVM
2022-06-24 11:10:35.774661 D | exec: Traceback (most recent call last):
2022-06-24 11:10:35.774827 D | exec:   File "/usr/sbin/ceph-volume", line 11, in <module>
2022-06-24 11:10:35.774966 D | exec:     load_entry_point('ceph-volume==1.0.0', 'console_scripts', 'ceph-volume')()
2022-06-24 11:10:35.775072 D | exec:   File "/usr/lib/python3.6/site-packages/ceph_volume/main.py", line 41, in __init__
2022-06-24 11:10:35.775220 D | exec:     self.main(self.argv)
2022-06-24 11:10:35.775721 D | exec:   File "/usr/lib/python3.6/site-packages/ceph_volume/decorators.py", line 59, in newfunc
2022-06-24 11:10:35.776151 D | exec:     return f(*a, **kw)
2022-06-24 11:10:35.776824 D | exec:   File "/usr/lib/python3.6/site-packages/ceph_volume/main.py", line 153, in main
2022-06-24 11:10:35.776835 D | exec:     terminal.dispatch(self.mapper, subcommand_args)
2022-06-24 11:10:35.776859 D | exec:   File "/usr/lib/python3.6/site-packages/ceph_volume/terminal.py", line 194, in dispatch
2022-06-24 11:10:35.776865 D | exec:     instance.main()
2022-06-24 11:10:35.776869 D | exec:   File "/usr/lib/python3.6/site-packages/ceph_volume/devices/lvm/main.py", line 46, in main
2022-06-24 11:10:35.776873 D | exec:     terminal.dispatch(self.mapper, self.argv)
2022-06-24 11:10:35.776877 D | exec:   File "/usr/lib/python3.6/site-packages/ceph_volume/terminal.py", line 194, in dispatch
2022-06-24 11:10:35.776881 D | exec:     instance.main()
2022-06-24 11:10:35.776886 D | exec:   File "/usr/lib/python3.6/site-packages/ceph_volume/decorators.py", line 16, in is_root
2022-06-24 11:10:35.776898 D | exec:     return func(*a, **kw)
2022-06-24 11:10:35.776902 D | exec:   File "/usr/lib/python3.6/site-packages/ceph_volume/devices/lvm/batch.py", line 431, in main
2022-06-24 11:10:35.776908 D | exec:     plan = self.get_plan(self.args)
2022-06-24 11:10:35.776912 D | exec:   File "/usr/lib/python3.6/site-packages/ceph_volume/devices/lvm/batch.py", line 469, in get_plan
2022-06-24 11:10:35.776915 D | exec:     args.wal_devices)
2022-06-24 11:10:35.776919 D | exec:   File "/usr/lib/python3.6/site-packages/ceph_volume/devices/lvm/batch.py", line 501, in get_deployment_layout
2022-06-24 11:10:35.776923 D | exec:     fast_type)
2022-06-24 11:10:35.776927 D | exec:   File "/usr/lib/python3.6/site-packages/ceph_volume/devices/lvm/batch.py", line 538, in fast_allocations
2022-06-24 11:10:35.776930 D | exec:     ret.extend(get_lvm_fast_allocs(lvm_devs))
2022-06-24 11:10:35.776934 D | exec:   File "/usr/lib/python3.6/site-packages/ceph_volume/devices/lvm/batch.py", line 167, in get_lvm_fast_allocs
2022-06-24 11:10:35.776939 D | exec:     disk.Size(b=int(d.lvs[0].lv_size)), 1) for d in lvs if not
2022-06-24 11:10:35.776943 D | exec:   File "/usr/lib/python3.6/site-packages/ceph_volume/devices/lvm/batch.py", line 168, in <listcomp>
2022-06-24 11:10:35.776946 D | exec:     d.used_by_ceph]
2022-06-24 11:10:35.776950 D | exec: IndexError: list index out of range
2022-06-24 11:10:35.816[369](https://github.com/rook/rook/runs/7040396677?check_suite_focus=true#step:8:370) C | rookcmd: failed to configure devices: failed to initialize osd: failed ceph-volume report: exit status 1
```

This is normal logs snippet:
```
2022-06-27 02:50:34.106460 I | cephosd: using /dev/test-rook-vg/test-rook-lv as metadataDevice for device /dev/sdb and let ceph-volume lvm batch decide how to create volumes
2022-06-27 02:50:34.106556 D | exec: Running command: stdbuf -oL ceph-volume --log-path /tmp/ceph-log lvm batch --prepare --bluestore --yes --osds-per-device 1 --crush-device-class hdd /dev/sdb --db-devices /dev/test-rook-vg/test-rook-lv --report
```

**Comment by BlaineEXE** [comment]
> @BlaineEXE I have resolved the comments besides integrated test of [osd-with-metadata-device](https://github.com/rook/rook/runs/7040396677?check_suite_focus=true#logs). I have another questions, why `ceph-volume --report` command runs twice, it seems that run it once should be enough. Any comments is welcome.
> 
> https://github.com/rook/rook/blob/c8658331df09cbef3b07b1ac3f60b28267ce1635/pkg/daemon/ceph/osd/volume.go#L752
> 
> 
> https://github.com/rook/rook/blob/c8658331df09cbef3b07b1ac3f60b28267ce1635/pkg/daemon/ceph/osd/volume.go#L761

I'm not certain, but I believe the first one may be intended to show the output in the logs for easy parsing during debugging. The second is what Rook uses to parse the output since it is in machine readable JSON.

File: pkg/daemon/ceph/osd/volume_test.go [hunk 3]
```diff
@@ -862,26 +868,244 @@ func TestInitializeBlock(t *testing.T) {
 			}
 
 			// First command
-			if args[9] == "--osds-per-device" && args[10] == "1" && args[11] == "/dev/sda" && args[12] == "--db-devices" && args[13] == "/dev/disk/by-id/nvme-Samsung_SSD_970_EVO_Plus_1TB_XXX" {
-				return `{"vg": {"devices": "/dev/disk/by-id/nvme-Samsung_SSD_970_EVO_Plus_1TB_XXX"}}`, nil
+			if args[9] == "--osds-per-device" && args[10] == "1" && args[11] == "/dev/sda" && args[12] == "--db-devices" && args[13] == "/dev/nvme0n1" {
+				return `{"vg": {"devices": "/dev/nvme0n1"}}`, nil
```
Comment (BlaineEXE) [inline pkg/daemon/ceph/osd/volume_test.go:872]: When modifying unit tests, it would be best to make a new test. We shouldn't be removing an existing test unless the behavior is changing, and the existing behavior shouldn't be changing, because we need to maintain backwards compatibility. I might suggest copying this test and making these changes in the new test.

**Review by BlaineEXE — CHANGES_REQUESTED** [review]

File: pkg/daemon/ceph/osd/volume_test.go (hunk 3 above)
Comment (microyahoo) [reply pkg/daemon/ceph/osd/volume_test.go:872]: The reason for the modification here is, the parameter `--db-devices` in  `ceph-volume` will use the device path name, but it will not affect the usage of the upper layer.

**Comment by microyahoo** [comment]
canary integration tests failed due to timeout

File: pkg/daemon/ceph/osd/volume.go [hunk 4]
```diff
@@ -609,6 +609,27 @@ func (a *OsdAgent) initializeDevicesLVMMode(context *clusterd.Context, devices *
...
+					}
+				}
+				if metadataDevice == nil {
+					return errors.Errorf("metadata device %s is not found", md)
+				}
+				// lvm device format is /dev/<vg>/<lv>
+				if metadataDevice.Type != sys.LVMType {
+					md = metadataDevice.Name
+				}
+
```
Comment (BlaineEXE) [inline pkg/daemon/ceph/osd/volume.go:632]: It looks like this also adds support for LVM devices as metadata.

@satoru-takeuchi does this look like a good implementation of that?

**Comment by BlaineEXE** [comment]
I think this looks good, but I want to get Satoru's opinion on the part related to LVM.

**Comment by satoru-takeuchi** [comment]
@microyahoo The OSD creation code path was changed in #7967 which was merged today. Could you rebase to the master and re-run CI?

**Comment by microyahoo** [comment]
> @microyahoo The OSD creation code path was changed in #7967 which was merged today. Could you rebase to the master and re-run CI?

Done. PTAL, thanks. @satoru-takeuchi @BlaineEXE

**Comment by microyahoo** [comment]
https://github.com/rook/rook/pull/10494#issuecomment-1166836151
@satoru-takeuchi I have another question, does the ceph layer support specifying `dm devices` as metadata devices like `--db-devices /dev/dm-0`  now?

**Comment by microyahoo** [comment]
The integration tests failed due to timeout, but I have no permission to retrigger it.

**Comment by parth-gr** [comment]
> The integration tests failed due to timeout, but I have no permission to retrigger it.

I re-trigger for you,
@travisn looks like a nice point for a self-bot if it can re-trigger CI for upstream contributors.

**Comment by BlaineEXE** [comment]
I think this still looks good. I'd like to get @satoru-takeuchi 's final sign off on it though.

@satoru-takeuchi would it be good to amend the `lvm` canary test to also test an LVM metadata device? Maybe that would be something you could do as a follow-on PR?

**Review by BlaineEXE — APPROVED** [review]
LGTM. @subhamkrai would you like to look again as well?

**Comment by satoru-takeuchi** [comment]
>  I have another question, does the ceph layer support specifying dm devices as metadata devices like --db-devices /dev/dm-0 now?

About logical volume, no. We should specify vg/lv style. `ceph-volume lvm batch` supports one more dm device, mpath. We can specify device name in this case.

I recalled that Rook can't use lvm as a metadata device. It's because that both host and osd container may use lv and vg at the same time. Then osd metadata may corrupt like the following document.

https://github.com/rook/rook/blob/master/Documentation/Troubleshooting/ceph-common-issues.md#lvm-metadata-can-be-corrupted-with-osd-on-lv-backed-pvc

So please remove logical volume support from this PR.

**Comment by satoru-takeuchi** [comment]
In addition, please update the document to show the lack of logical volume support.

https://github.com/rook/rook/blob/master/Documentation/CRDs/Cluster/ceph-cluster-crd.md#osd-configuration-settings

**Comment by microyahoo** [comment]
> > I have another question, does the ceph layer support specifying dm devices as metadata devices like --db-devices /dev/dm-0 now?
> 
> About logical volume, no. We should specify vg/lv style. `ceph-volume lvm batch` supports one more dm device, mpath. We can specify device name in this case.
> 
> I recalled that Rook can't use lvm as a metadata device. It's because that both host and osd container may use lv and vg at the same time. Then osd metadata may corrupt like the following document.
> 
> https://github.com/rook/rook/blob/master/Documentation/Troubleshooting/ceph-common-issues.md#lvm-metadata-can-be-corrupted-with-osd-on-lv-backed-pvc
> 
> So please remove logical volume support from this PR.

Thanks for your reminder, I have to make a clarification. 
Canary integration tests [osd-with-metadata-device](https://github.com/rook/rook/runs/7307087691?check_suite_focus=true#logs) uses `/dev/vg/lv` as a metadata device, its device name format is `/dev/mapper/xxx`. Therefore, I added a judgment, if the user specifies a metadata device name with lvm type, the device name is not used, but the user specified `/dev/vg/lv` is used.
```
🍺 /root/go/src/rook/rook ☞ git:(issue_10486)lvs -a -o +devices    
File descriptor 12 (pipe:[1699532284]) leaked on lvs invocation. Parent PID 31828: zsh
  LV            VG            Attr       LSize   Pool Origin Data%  Meta%  Move Log Cpy%Sync Convert Devices         
  root          centos        -wi-ao----  92.99g                                                     /dev/vda2(0)    
  root          centos        -wi-ao----  92.99g                                                     /dev/vda3(0)    
  swap          centos        -wi-a-----   6.00g                                                     /dev/vda2(13567)
  test-rook-lv  test-rook-vg  -wi-ao---- <10.00g                                                     /dev/loop0(0)   
  test-rook-lv1 test-rook-vg1 -wi-a----- <10.00g                                                     /dev/loop1(0)   
  test-rook-lv2 test-rook-vg2 -wi-a----- <10.00g                                                     /dev/loop2(0)   
🍺 /root/go/src/rook/rook ☞ git:(issue_10486)ll /dev/test-rook-vg/test-rook-lv
lrwxrwxrwx 1 root 7 Jun 28 14:45 /dev/test-rook-vg/test-rook-lv -> ../dm-2
🍺 /root/go/src/rook/rook ☞ git:(issue_10486)ll /dev/test-rook-vg1/test-rook-lv1 
lrwxrwxrwx 1 root 7 Jun 27 15:30 /dev/test-rook-vg1/test-rook-lv1 -> ../dm-3

🍺 /root/go/src/rook/rook ☞ git:(issue_10486)lsblk /dev/test-rook-vg1/test-rook-lv1 --bytes --nodeps --pairs --paths --output SIZE,ROTA,RO,TYPE,PKNAME,NAME,KNAME,MOUNTPOINT,FSTYPE
SIZE="10733223936" ROTA="1" RO="0" TYPE="lvm" PKNAME="" NAME="/dev/mapper/test--rook--vg1-test--rook--lv1" KNAME="/dev/dm-3" MOUNTPOINT="" FSTYPE=""

```

**Comment by satoru-takeuchi** [comment]
> Canary integration tests osd-with-metadata-device uses /dev/vg/lv as a metadata device, its device name format is /dev/mapper/xxx. Therefore, I added a judgment, if the user specifies a metadata device name with lvm type, the device name is not used, but the user specified /dev/vg/lv is used.

So the following description was wrong. I apologize to you.

> We should specify vg/lv style.

I added `osd-with-metadata-device` canary test. I shouldn't use logical volume in this case. I used logical volume at that time because CI environment has only one device, but partition can' be used for metadata device.

One more problem is that some users may have already used logical volume as metadata. It's a regression for such users to remove logical volume support for metadata device. 

Please give me a little more time to find another solution.

**Comment by microyahoo** [comment]
> I added `osd-with-metadata-device` canary test. I shouldn't use logical volume in this case. I used logical volume at that time because CI environment has only one device, but partition can' be used for metadata device.
> 
IMO, It is feasible to use lvm as a metadata disk to test, except for the problem you mentioned.

**Comment by microyahoo** [comment]
> In addition, please update the document to show the lack of logical volume support.
> 
> https://github.com/rook/rook/blob/master/Documentation/CRDs/Cluster/ceph-cluster-crd.md#osd-configuration-settings

I will update document later.

**Comment by satoru-takeuchi** [comment]
@microyahoo It seems that my concern was overblown. It's OK to specify an LV as a metadata device. OSD data corruption, which I described, happened because both host and osd container access lvm tags. However, Ceph doesn't access lvm tags of metadata devices.

So my change request is just to clarify the supported devices (devices and LVs). Other changes LGTM.

**Review by satoru-takeuchi — APPROVED** [review]

**Review by satoru-takeuchi — APPROVED** [review]

**Comment by satoru-takeuchi** [comment]
@subhamkrai PTAL whether your comment was reflected.

**Review by subhamkrai — APPROVED** [review]
